Allow specifying SIP username for outgoing calls

jaminh opened the pull request:
Allow configuring a SIP username to be sent in outgoing call requests to identify the home assistant source endpoint.

<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Allow configuring a SIP username to be sent in outgoing call requests to identify the home assistant source endpoint.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 
- Link to developer documentation pull request: 
- Link to frontend pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [ ] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

jaminh committed:
Allow specifying SIP username for outgoing calls

Allow configuring a SIP username to be sent in outgoing call requests to
identify the home assistant source endpoint.

jaminh commented:
Updated documentation for change https://github.com/home-assistant/home-assistant.io/pull/37264

balloob commented on homeassistant/components/voip/const.py:
```diff
@@ -12,4 +12,7 @@
     "sleep_ratio": 0.99,
 }
 
+DEFAULT_SIP_USER = "HA"
```
Could this impact how systems respond? Why not make the user optional if it's not set.

balloob commented on homeassistant/components/voip/config_flow.py:
```diff
@@ -70,7 +73,35 @@ async def async_step_init(
                         ),
-                    ): cv.port
+                    ): cv.port,
+                    vol.Optional("enable_advanced"): bool,
```
Lets do it without advanced checkbox. Guess its all advanced anyway.

jaminh committed:
Remove advanced options section

balloob commented on tests/components/voip/test_config_flow.py:
```diff
@@ -79,4 +79,15 @@ async def test_options_flow(hass: HomeAssistant) -> None:
+        user_input={"sip_port": 5061, "sip_user": "HA"},
+    )
+    assert result["type"] is FlowResultType.CREATE_ENTRY
+    assert config_entry.options == {"sip_port": 5061, "sip_user": "HA"}
```
Can you add a test to remove it too.

balloob reviewed (changes requested):
Please add the reverse test.

jaminh committed:
Add test for removing user

jaminh commented:
I suspect there might be an issue with user getting set to an empty string rather than None.

jaminh committed:
Allow unsetting SIP user

Make previous SIP user value a suggested value rather than default to
allow unsetting by submitting an empty value in the form.

balloob commented on homeassistant/components/voip/config_flow.py:
```diff
@@ -36,6 +41,10 @@ async def async_step_user(
                 step_id="user",
             )
 
+        if CONF_SIP_USER in user_input and is_none_or_empty(
```
This is not necessary for the main flow, as we don't ask for any values. The user is only set in the options flow.

balloob commented on homeassistant/components/voip/config_flow.py:
```diff
@@ -58,7 +67,17 @@ async def async_step_init(
         """Manage the options."""
         if user_input is not None:
-            return self.async_create_entry(title="", data=user_input)
+            if CONF_SIP_USER in user_input and is_none_or_empty(
+                user_input.get(CONF_SIP_USER, "")
+            ):
```
```suggestion
            if CONF_SIP_USER in user_input and not user_input[CONF_SIP_USER]:
```

jaminh committed:
Remove unnecessary checks

Remove user check from main flow and remove none or empty check.
